### ISSUE TITLE
Add toggle for nvim-cmp

### ISSUE THREAD

**Issue by osamuaoki** [LazyVim/LazyVim#2981]
### Did you check the docs?

- [X] I have read all the LazyVim docs

### Is your feature request related to a problem? Please describe.

When modifying an existing code and I know what exactly to be typed, auto completion by nvim-cmp may interfare with my  typing just as much as auto-pair feature.  Although you have added toggle feature to mini.pairs using `vim.g.minipairs_disable`, you haven't made similar changes to nvim-cmp.

### Describe the solution you'd like

Change `lua/lazyvim/plugins/codeng.lua` to accommodate `vim.g.cmp_disable` by redefining the function `enabeled` along what I am doing in my `~/.config/nvim/lua/plugins/coding.lua`:

```
  -- auto completion
return {
  {
    "hrsh7th/nvim-cmp",
    opts = function()
      require("cmp").setup {
        -- * https://github.com/hrsh7th/nvim-cmp/issues/429
        -- * https://github.com/hrsh7th/nvim-cmp/issues/261
        -- function source copied from https://github.com/hrsh7th/nvim-cmp.git and modified
        -- path: nvim-cmp/lua/cmp/config/default.lua , section: enabled = function() ... end
        enabled = function()
          --if vim.fn.exists("b:cmp_disable") == 0 then
          local disabled = false
          disabled = disabled or (vim.api.nvim_get_option_value("buftype", {buf=0}) == "prompt")
          disabled = disabled or (vim.fn.reg_recording() ~= '')
          disabled = disabled or (vim.fn.reg_executing() ~= '')
          -- the following is the modification to accomodate global auto completion toggle
          disabled = disabled or vim.g.cmp_disable
          -- if any one of the above is true, return false / otherwise return false
          return not disabled
        end,
      }
    end,
  },
}
```
Along with equivalent keymap definition in `lua/lazyvim/config/keymap.lua` like my local one under `~/.config`:
```
map("n", "<leader>um", function()
  vim.g.cmp_disable = not vim.g.cmp_disable
  if vim.g.cmp_disable then
    vim.notify("Disabled CMP", vim.log.levels.INFO, { title = "CMP" })
  else
    vim.notify("Enabled CMP", vim.log.levels.INFO, { title = "CMP" })
  end
end, { desc = "Toggle CMP †" })
```
Here, `vim.notify` may be replaced with LazyVim equivalent.

I am using nvim 0.10 but if you want to address older release, 

```
          disabled = disabled or (vim.api.nvim_get_option_value("buftype", {buf=0}) == "prompt")
```
may be replaced by : 
```  
        disabled = disabled or (vim.api.nvim_buf_get_option(0, 'buftype') == 'prompt')
 ```

In addition to the above, adding helpful entry to ``lua/lazyvim/config/options.lua` such as the following may be a good idea.

```
-- Input Pairs '' "" () [] {} `` (toggle with "<leader>up")
--   * true:  Disable
--   * false: Enable
vim.g.minipairs_disable = true
-- vim.g.minipairs_disable = false

-- CMP (toggle with "<leader>um")
--   * true: Disable
--   * false:  Enable
vim.g.cmp_disable = true
-- vim.g.cmp_disable = false
```

### Describe alternatives you've considered

I thought about toggle for each buffer, too.  But unlike auto format, I thought cluttering which-key UI is not worth complication.

### Additional context

FYI:
I think indentation related feature also deserves toggle.  `mini.indentscope` is easy but I still need to figure out `indent-blankline.nvim` to enable copy from terminal using mouse select (not by visual mode).

**Comment by dpetka2001** [comment]
Why don't you just disable auto-completion and call it with `<C-Space>` whenever you need it? This seems more like a personal preference to me.

**Comment by osamuaoki** [comment]
Does `<C-SPACE>` stops nvim-cmp persistently?  I don't want to see visual noise so I don't want to type `<C-SPACE>` or whatever (like `<C-e>`) before `ENTER`.

Solution proposed here is to have quiet persistent work environment when one needs it easily while keeping normal UI for the rest of time.

I do use `<C-e>` to ignore nvim-cmp suggestions when I need.  But that's when I want to see such suggestions.

FYI: I usually have `<C-SPACE>` mapped for input method. 

Are you talking mmapping like one in https://github.com/neovim/nvim-lspconfig/wiki/Autocompletion#nvim-cmp ?  I was annoyed for snipped suggestions (not LSP).

**Comment by dpetka2001** [comment]
I'm talking about [this setting](https://github.com/hrsh7th/nvim-cmp/blob/ce16de5665c766f39c271705b17fff06f7bcb84f/doc/cmp.txt#L506-L509). That way you'll never get auto-completion popup menu unless you invoke it explicitly with `<C-Space>` (which is the keymap for `cmp.mapping.complete()` or whatever other keymap you use for that if you have changed it in your personal configuration).

**Comment by osamuaoki** [comment]
Good point, I see.  Using [completion.autocomplete](https://github.com/hrsh7th/nvim-cmp/blob/ce16de5665c766f39c271705b17fff06f7bcb84f/doc/cmp.txt#L506-L509) is certainly a cleaner approach to address this problem.

My main issue is this functionality is not readily accessible like "`<SPACE>up ➜ Toggle Auto Pairs`" and all similar toggles in which-key menu under "`<space>u`" menu.

**Comment by dpetka2001** [comment]
With that setting it doesn't have to be toggleable. The user just presses `<C-Space>` whenever he wants the auto-completion menu to pop up.

**Comment by osamuaoki** [comment]
I admit that had very weak understanding of how plugin configuration is working.  Thank you for your guidance.  I still have issues. Please be paitient with me.

I removed all my local changes (including OS level `<C-SPACE>` usage) and tried what you said since it sounded good. 
 But I am not getting what I expected.   Here is my situation as shell.
```sh
cat > test.txt << EOF
abc
ac1
ac2
accc
adddd
EOF
nvim '+$' test.txt
```
This puts me into nvim and cursor at the last line.  I then type "o" and "a" and "`<C-SPACE>`".  Nothing happens.

![Screenshot from 2024-04-19 10-51-08](https://github.com/LazyVim/LazyVim/assets/7339387/c2230a76-cfde-4705-9b68-c4e3c2e84a14)

Since I have no local changes to [the default LazyVim plugin setting](https://github.com/LazyVim/LazyVim/blob/bb36f71b77d8e15788a5b62c82a1c9ec7b209e49/lua/lazyvim/plugins/coding.lua#L32-L48), I am wondering why.

Here, I see following working fine:
* "`<C-e>`"
* "`<CR>`"
* "`<C-n>`"
* "`<C-p>`"

But the followings happened.
* "`<C-SPACE>`" -- Do nothing
* "`<C-b>`" -- just insert control code "^B" (0x02).
* "`<C-f>`" -- - just insert control code "^F" (0x06). 

Is there some combination issues for associated snippet lua code?

"`<C-b>`"  and "`<C-f>`" may be OK but non-functioning "`<C-SPACE>`" is problematic.

FYI: I used  locally compiled nvim with 0.9 and 0.10 versions for testing.  In order to make sure the particular key choice of "`<C-SPACE>`" is conflicting, I also tested with changed key binding but it didn't work.

**Comment by dpetka2001** [comment]
`<C-Space>` is actually invoking manually the popup auto-completion menu. It doesn't do anything else. What are you expecting it to do? It doesn't do anything when the popup auto-completion menu is open.

**Comment by osamuaoki** [comment]
What I expected was to see some kind of selection menu pop up selection when I press `<C-Space>` from you comment.

Yes, I now can see manual pop up selection with `<C-SPACE>` but the current selection doesn't have such item.
![Screenshot from 2024-04-19 21-50-11](https://github.com/LazyVim/LazyVim/assets/7339387/c5b16860-e602-4009-baef-0a34ee5d5f47)

I am not sure how and where your suggested configuration can be added to realize enable/disable auto completion.

(As for `<C-f>/<C-b>`, your comment make sense and matches my guess of situation.)

**Comment by dpetka2001** [comment]
From your own screenshot

![2024-04-19_17-21](https://github.com/LazyVim/LazyVim/assets/12776461/0517f556-347c-49d4-8094-e2075c5cc9ba)

What you see inside the red rectangle is the auto-completion menu. This is what `<C-space>` opens. Just put the code snippet in my previous post in a new file in `~/.config/nvim/lua/plugins/` directory and that menu won't open unless you specifically press `<C-space>`.

**Comment by osamuaoki** [comment]
Thanks for your help.  Now I know better way to configure nvim-cmp.

In light of my new understanding, let me update description of my wish.

* Add key binding to toggle between current default mode and `autocomplete = false` dynamically. (similar to what `<LEADER>up`.)
* Add a single global variable like `vim.g.minipairs_diable` offered by mini.pairs to enable easy setting of starting mode.

I may try to update my proposed script later. (I will be off-line for a month.  So don't expect my quick response, please.)

**Comment by dpetka2001** [comment]
I don't believe there's a reason to make something toggleable if there's already an inherent option that allows the user to call auto-completion at will.

**Comment by osamuaoki** [comment]
It's OK to be "not planned" :-)

This discussion lead me to set up "Toggle for auto/manual pop-up for autocomplete" locally as below with which-key v3.

`~/.config/nvim/lua/plugins/nvim-cmp.lua`:
```
-- Toggle autocomplete vs. <C-SPACE> complete
return {
  -- disable autocomplete if vim.g.cmp_disable is true
  {
    "hrsh7th/nvim-cmp",
    opts = function()
      require("cmp").setup({
        completion = { autocomplete = not vim.g.autocomplete_disable }
      })
    end,
  },
}
```
`~/.config/nvim/lua/config/keymaps.lua` contains:
```
-- Toggle Auto-popup complete vs. manual <C-SPACE> complete
local cmp = require("cmp")
LazyVim.toggle.map(
  "<leader>ua",
  LazyVim.toggle.wrap({
    name = "Autocomplete",
    get = function()
      return cmp.get_config().completion.autocomplete == false
    end,
    set = function(state)
      if state then
        cmp.setup({ completion = { autocomplete = false } })
      else
        cmp.setup({ completion = { autocomplete = { cmp.TriggerEvent.TextChanged } } })
      end
    end,
  })
)
```

`~/.config/nvim/lua/config/options.lua` contains:
```
vim.g.autocomplete_disable = true -- lua/plugins/nvim-cmp.lua
```
